src/core/ObservableQuery.ts

```diff
@@ -979,6 +989,15 @@ Did you mean to call refetch(variables) instead of refetch({ variables })?`,
     const oldFetchPolicy = this.options.fetchPolicy;
 
     const mergedOptions = compact(this.options, newOptions || {});
+
+    // Allow variables to be reset back to `undefined` if explicitly passed as
+    // variables: undefined to reobserve, otherwise `compact` will ignore the
+    // `variables` key from from `newOptions`.
+    mergedOptions.variables =
+      newOptions && hasOwnProperty.call(newOptions, "variables") ?
```

Comment:
```suggestion
      newOptions && "variables" in newOptions ?
```
I think this would do it, too - or do we explicitly want to guard against objects that have inherited `variables` via the prototype chain?